File: 1074-number-of-submatrices-that-sum-to-target/1074-number-of-submatrices-that-sum-to-target.py
class Solution(object):
    def numSubmatrixSumTarget(self, mat, target):
        n, m = len(mat), len(mat[0])
        for i in range(n):
            for j in range(1, m):
                mat[i][j] += mat[i][j-1]
        
        ans = 0
        for i in range(m):
            for col in range(i, m):
                mp, _sum = {}, 0
                mp[0] = 1
                for row in range(n):
                    _sum += mat[row][col]
                    _sum -= mat[row][i-1] if i > 0 else 0
                    ans += mp.get(_sum - target, 0)
                    mp[_sum] = mp.get(_sum, 0) + 1
        
        return ans


                
        

File: 1074-number-of-submatrices-that-sum-to-target/1074-number-of-submatrices-that-sum-to-target.cpp
class Solution {
public:
    int numSubmatrixSumTarget(vector<vector<int>>& mat, int target) {
        int n = mat.size(), m = mat[0].size();
        for(int i = 0; i<n; i++) {
            for(int j = 1; j<m; j++) {
                mat[i][j] += mat[i][j-1];
            }
        }

        int ans = 0;
        for(int i = 0; i<m; i++) {
            for(int col = i; col<m; col++) {
                unordered_map<int, int> mp;
                mp[0]++;
                int sum = 0;
                for(int row = 0; row<n; row++) {
                    sum += mat[row][col] - (i > 0 ? mat[row][i-1] : 0);
                    ans += mp[sum - target];
                    mp[sum]++;
                }
            }
        }

        return ans;
    }
};